src/handler/sign_tx.c

```diff
@@ -241,10 +243,17 @@ bool _decode_elements() {
             // still decoding tokens, but this one was divided between calls
             THROW(TX_STATE_READY);
         }
-        // Safely discard data
+        uint8_t uid[TOKEN_UID_LEN] = {0};
+        memmove(uid, G_context.tx_info.buffer, TOKEN_UID_LEN);
```

Comment:
Just to make sure I understand:

- `G_context` is a global variable
- `uid` is allocated on the stack, for the 32-byte sequence of the token-uid (which is also the tx_hash of the token_creation_tx of the token)
- `memmove` is used to copy the uid which is currently accessible from `G_context.tx_info.buffer`

Throughout the code we seem to use `memmove` for a lot of plain copies. Is `memcpy` not available? Or is there any particular reason we _have_ to use `memmove`. I'm not really familiar with the implications of the hardware restrictions, maybe this has to do with that. Because my first choice would be to use `memcpy` unless we're moving bytes along the same array. I guess there is no harm though.

Another observation, do we really need to copy these bytes? Wouldn't it be enough for `uid` to be a pointer cast from `G_context.tx_info.buffer` instead of being allocated on the stack?

Reply:
> `G_context` is a global variable

Yes, the context saves the data ledger receives and we read this data, decoding each element of the transaction, in this line the token uid array.

> uid is allocated on the stack, for the 32-byte sequence of the token-uid (which is also the tx_hash of the token_creation_tx of the token)

Yes, we need this uid to check if the token was verified with `SEND_TOKEN_DATA` (i.e. it's present on the registry).

> Is `memcpy` not available?

It is, but since the operations do the same but `memmove` is safer when copying data on the same buffer the default I use is `memmove`. But we could safely use `memcpy` here for instance, but not below where we use memmove to "move" the buffer by 32 bytes of the uid we read.

Reply:
Got it!

But what about this:

> Wouldn't it be enough for uid to be a pointer cast from G_context.tx_info.buffer instead of being allocated on the stack?

Couldn't we use a pointer directly to the buffer instead of copying?

It's only 32 bytes, but that could be a lot, and if we don't really need it, it's free memory estate.